plugin/cache/cache.go

```diff
@@ -201,23 +207,31 @@ func (w *ResponseWriter) WriteMsg(res *dns.Msg) error {
 	return w.ResponseWriter.WriteMsg(res)
 }
 
-func (w *ResponseWriter) set(m *dns.Msg, key uint64, mt response.Type, duration time.Duration) {
-	// duration is expected > 0
-	// and key is valid
+func cacheType(mt response.Type) int8 {
 	switch mt {
 	case response.NoError, response.Delegation:
-		i := newItem(m, w.now(), duration)
-		w.pcache.Add(key, i)
-
+		return posCache
 	case response.NameError, response.NoData, response.ServerError:
-		i := newItem(m, w.now(), duration)
-		w.ncache.Add(key, i)
-
+		return negCache
 	case response.OtherError:
-		// don't cache these
+		return noCache
```

Comment:
Please remove this indirection; it's not needed and it bolts an indirection on an indirection